pkg/sql/pgwire/pgwirebase/too_big_error.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package pgwirebase
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/cockroachdb/errors"
+	"github.com/gogo/protobuf/proto"
+)
+
+// withMessageTooBig decorates an error when a read overflows the ReadBuffer.
+type withMessageTooBig struct {
+	cause error
+	size  int
+}
+
+var _ error = (*withMessageTooBig)(nil)
+var _ errors.SafeDetailer = (*withMessageTooBig)(nil)
+var _ fmt.Formatter = (*withMessageTooBig)(nil)
+var _ errors.Formatter = (*withMessageTooBig)(nil)
+
+func (w *withMessageTooBig) Error() string         { return w.cause.Error() }
+func (w *withMessageTooBig) Cause() error          { return w.cause }
+func (w *withMessageTooBig) Unwrap() error         { return w.cause }
+func (w *withMessageTooBig) SafeDetails() []string { return []string{strconv.Itoa(w.size)} }
+
+func (w *withMessageTooBig) Format(s fmt.State, verb rune) { errors.FormatError(w, s, verb) }
+
+func (w *withMessageTooBig) FormatError(p errors.Printer) (next error) {
+	return w.cause
+}
+
+// decodeWithMessageTooBig is a custom decoder that will be used when decoding
+// withMessageTooBig error objects.
+// Note that as the last argument it takes proto.Message (and not
+// protoutil.Message which is required by linter) because the latter brings in
+// additional dependencies into this package and the former is sufficient here.
+func decodeWithMessageTooBig(
+	_ context.Context, cause error, _ string, details []string, _ proto.Message,
+) error {
+	size := -1
+	if len(details) > 0 {
+		var err error
+		size, err = strconv.Atoi(details[0])
+		if err != nil {
+			return err
```

Comment:
it's an API violation to return an error which is not the result of decoding.

You have two options:
- either decode fails, in which case you must return `nil` (the errors library will do the right thing and still decode _something_ that's alike to your original object)
- or actually return a `&withMessageTooBig{}`.

Reply:
ah, done!